apis/rust/operator/src/raw.rs

```diff
@@ -32,18 +23,14 @@ pub unsafe fn dora_drop_operator<O>(operator_context: *mut c_void) {
 }
 
 pub unsafe fn dora_on_input<O: DoraOperator>(
-    id_start: *const u8,
-    id_len: usize,
+    metadata: *const c_void,
     data_start: *const u8,
     data_len: usize,
     output_fn_raw: OutputFnRaw,
     dora_context: *const c_void,
     operator_context: *mut c_void,
 ) -> isize {
-    let id = match std::str::from_utf8(unsafe { slice::from_raw_parts(id_start, id_len) }) {
-        Ok(id) => id,
-        Err(_) => return -1,
-    };
+    let metadata = unsafe { &*metadata.cast() };
```

Comment:
`Metadata` is not `repr(C)`, so it's data layout might change between compiler versions. So this cast can lead to undefined behavior if the operator was compiled with a different compiler version than the runtime.

Reply:
I see. So, but if we either remove or change the `id`field, we can tag the struct `#[repr(C)]` and we would be able to pass the data around.

Reply:
Unfortunately, adding `#[repr(C)]` to `Metadata` is not enough because `String` itself is not FFI-safe. See this example: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b20f8605352f30e25e664189ea6486db

```rust
#[repr(C)]
struct Foo {
    field: String,
}

extern "C" fn _assert_ffi(_v: Foo) {}
```

The compiler throws this warning:

```
warning: `extern` fn uses type `String`, which is not FFI-safe
 --> src/main.rs:6:31
  |
6 | extern "C" fn _assert_ffi(_v: Foo) {}
  |                               ^^^ not FFI-safe
  |
  = note: `#[warn(improper_ctypes_definitions)]` on by default
  = help: consider adding a `#[repr(C)]` or `#[repr(transparent)]` attribute to this struct
  = note: this struct has unspecified layout
```

The only reason that we don't see such a warning for your code is that you pass a void pointer and then cast it again. But the issue still exists.

---

I'm currently trying to find a way to autogenerate these FFI bindings to make this easier. Until then, we have to pass strings etc as pointers + length tuples to be safe...